File: src/client/modules/Controllers/MovementController.lua
local PRINT_DIAGNOSES = false
local PRINT_STARTS = false
local PRINT_STOPS = false
local VISUALIZE = false

local BASE_CLIMB_SPEED = 10
local BASE_WALK_SPEED = 20
local DASH_DURATION = 0.4
local MAX_CLIMB_BONUS = 10 -- bonus for max climb training

local Players = game:GetService("Players")
local ReplicatedStorage = game:GetService("ReplicatedStorage")
local RunService = game:GetService("RunService")
local TweenService = game:GetService("TweenService")
local UserInputService = game:GetService("UserInputService")

local Knit = require(ReplicatedStorage.Packages.Knit)
local Trove = require(ReplicatedStorage.Packages.Trove)

local ActionService, IdentityService, ManaService
local AnimationController, InputController, MagicController

local CastVisualizer = require(ReplicatedStorage.Utility.CastVisualizer)

local MovementController = Knit.CreateController {
	Name = "MovementController"
}

local VFX = ReplicatedStorage.Effects.Visuals
local SFX = ReplicatedStorage.Effects.Sounds

local raycastParams = RaycastParams.new()
raycastParams.FilterType = Enum.RaycastFilterType.Exclude
raycastParams.CollisionGroup = "Characters"
raycastParams.IgnoreWater = true

function MovementController.HasLineOfSight(from : Vector3, to : Vector3, params : RaycastParams)
	local ray = workspace:Raycast(from, (to-from), params)
	return ray and ((to-from).Magnitude-ray.Distance) > 0.1
end

function MovementController:AlignCharacter(deltaTime : number, ray : RaycastResult)
	local hrp = self.Character:FindFirstChild("HumanoidRootPart")

	local newX = ray.Position.X + ray.Normal.X
	local newY = ray.Position.Y + ray.Normal.Y
	local newZ = ray.Position.Z + ray.Normal.Z
	local pos = Vector3.new(newX, newY, newZ)
	local lookAt = ray.Position

	local goal = CFrame.new(pos, lookAt)
	self.GoalPart.CFrame = goal

	hrp.CFrame = hrp.CFrame:Lerp(goal, 1 - 1e-9 ^ deltaTime)
end

function MovementController:ClimbStart(ray : RaycastResult)
	self._climbTrove = Trove.new()

	if PRINT_STARTS then print('start \t\tclimb') end

	MagicController:ChargeMana(false)

	local character = self.Character

	self.ClimbForce = self._climbTrove:Clone(script.ClimbForce)
	self.ClimbForce.Parent = character
	self.ClimbForce.Attachment0 = character.HumanoidRootPart.RootAttachment
	self.ClimbForce.MaxForce = math.huge

	self.GoalPart = self._climbTrove:Clone(script.GoalPart)
	self.GoalPart.Parent = character.HumanoidRootPart

	self.GoalAttachment = self._climbTrove:Add(Instance.new("Attachment"))
	self.GoalAttachment.Parent = self.GoalPart

	self.ClimbConstraint = self._climbTrove:Clone(script.ClimbConstraint)
	self.ClimbConstraint.Parent = character
	self.ClimbConstraint.Attachment0 = character.HumanoidRootPart.RootAttachment
	self.ClimbConstraint.Attachment1 = self.GoalAttachment

	local humanoid = character:FindFirstChild("Humanoid")
	humanoid.AutoRotate = false

	self:AlignCharacter(0, ray)

	humanoid:SetStateEnabled(Enum.HumanoidStateType.Freefall, false)
	humanoid:SetStateEnabled(Enum.HumanoidStateType.Running, false)
	humanoid:ChangeState(Enum.HumanoidStateType.Climbing)

	self._climbTrove:BindToRenderStep("climb_update", Enum.RenderPriority.Character.Value, function(deltaTime)
		self:ClimbUpdate(deltaTime)
	end)

	local input = Vector2.new(InputController.GetAxis("horizontal"), InputController.GetAxis("vertical"))
	if input.Y ~= 0 then
		AnimationController:Play("ClimbUp")
	elseif input.X == 1 then
		AnimationController:Play("ClimbRight")
	elseif input.X == -1 then
		AnimationController:Play("ClimbLeft")
	else
		AnimationController:Play("ClimbIdle")
	end

	self._climbTrove:Connect(UserInputService.InputBegan, function(...)
		AnimationController:HandleClimbAnims(...)
	end)
	self._climbTrove:Connect(UserInputService.InputEnded, function(...)
		AnimationController:HandleClimbAnims(...)
	end)

	self.Climbing = true
	ActionService.ActionStarted:Fire("climb")
end

function MovementController:ClimbStop()
	if PRINT_STOPS then print('stop \t\tclimb') end

	local humanoid = self.Character:FindFirstChild("Humanoid")
	humanoid.AutoRotate = true
	humanoid:SetStateEnabled(Enum.HumanoidStateType.Freefall, true)
	humanoid:SetStateEnabled(Enum.HumanoidStateType.Running, true)

	AnimationController:Stop("ClimbUp")
	AnimationController:Stop("ClimbDown")
	AnimationController:Stop("ClimbLeft")
	AnimationController:Stop("ClimbRight")
	AnimationController:Stop("ClimbIdle")

	game:GetService("ContextActionService"):UnbindAction("handle_climb_anims")

	self.Climbing = false
	self._climbTrove:Destroy()
	ActionService.ActionStopped:Fire("climb")
end

function MovementController:ClimbUp(goal : CFrame)
	if not self.Climbing then return end

	local hrp = self.Character.HumanoidRootPart
	local humanoid = self.Character.Humanoid

	local lookVector = hrp.CFrame.LookVector
	local uVector = hrp.CFrame.UpVector

	hrp.Anchored = true

	goal = goal + lookVector * 0.5 + uVector
	local tween = TweenService:Create(hrp, TweenInfo.new(.25, Enum.EasingStyle.Linear), {CFrame = goal})
	AnimationController:Play("LedgeClimbUp")
	tween:Play()
	self._trove:Connect(tween.Completed, function(playbackState)
		hrp.Anchored = false
		self:ClimbStop()
	end)
end

function MovementController:ClimbUpdate(deltaTime : number)
	if PRINT_DIAGNOSES then print("climb update") end
	local input = Vector2.new(InputController.GetAxis("horizontal"), InputController.GetAxis("vertical"))
	local hrp = self.Character:FindFirstChild("HumanoidRootPart")

	local sessionData = Knit.Player.Data

	local statChanges = sessionData.StatChanges
	local climbBoost = statChanges.ClimbBoost.Value

	local mana = sessionData.ManaAmount.Value
	local climbTrains = 0

	local CLIMB_SPEED = ((-1/1000) * mana * (mana-200) + BASE_CLIMB_SPEED + climbTrains/MAX_CLIMB_BONUS) * (1+climbBoost)

	local lookVector = hrp.CFrame.LookVector
	local rVector = hrp.CFrame.RightVector
	local uVector = hrp.CFrame.UpVector

	local bottomDwn = workspace:Raycast(hrp.Position + Vector3.new(0, -0.5, 0), -uVector, raycastParams)

	if bottomDwn then if PRINT_DIAGNOSES then warn("too close to ground") end self:ClimbStop() end

	local diagonalIn, diagonalOut, sideways
	if input.X ~= 0 then
		local inOrigin	= hrp.Position + rVector *  input.X
		local inPoint  	= hrp.Position + rVector * -input.X + lookVector * 2
		local outOrigin = hrp.Position + rVector * -input.X
		local outPoint 	= hrp.Position + rVector *  input.X + lookVector * 2

		diagonalIn = workspace:Raycast(inOrigin, inPoint-inOrigin, raycastParams)
		diagonalOut = workspace:Raycast(outOrigin, outPoint-outOrigin, raycastParams)
		sideways = workspace:Raycast(hrp.Position, input.X * rVector * 2, raycastParams)
	end
	if VISUALIZE then CastVisualizer.visualize(hrp.Position, hrp.Position+input.X * rVector * 1.5, raycastParams) end

	local blockSize = Vector3.new(hrp.Size.X, hrp.Size.Y/2, 0.1)
	local blockFwd = workspace:Blockcast(hrp.CFrame, blockSize, lookVector * 2, raycastParams)
	if VISUALIZE then CastVisualizer.visualizeBlock(hrp.CFrame, blockSize, lookVector * 2, raycastParams) end

	if not blockFwd and not diagonalIn and not diagonalOut then if PRINT_DIAGNOSES then warn("too far from surface") end self:ClimbStop() end

	local seesTop = false
	local headCastCFrame = hrp.CFrame + uVector * 1.5 - lookVector * 0.5
	local headCastSize = Vector3.new(2,0.1,0.1)
	if not workspace:Blockcast(headCastCFrame, headCastSize, lookVector * 2, raycastParams) then
		seesTop = MovementController.HasLineOfSight(headCastCFrame.Position, hrp.Position + lookVector * 1.5 + Vector3.new(0,1,0), raycastParams)
		if VISUALIZE then CastVisualizer.visualizeLineOfSight(headCastCFrame.Position, hrp.Position + lookVector * 1.5 + Vector3.new(0,1,0), raycastParams) end
	end

	local centerFwd = workspace:Raycast(hrp.Position, lookVector * 2, raycastParams)

	if seesTop then
		local ledgeBlockCFrame = hrp.CFrame * CFrame.Angles(math.rad(-10),0,0) * CFrame.new(0,1,0)
		local ledgeBlockSize = Vector3.new(0.01,2,0.01)
		local ledgeBlockCast = workspace:Blockcast(ledgeBlockCFrame, ledgeBlockSize, lookVector, raycastParams)

		if ledgeBlockCast then
			self:ClimbUp(CFrame.new(ledgeBlockCast.Position) * hrp.CFrame.Rotation)
		else
			if PRINT_DIAGNOSES then warn("ledge missing") end
			self:ClimbStop()
		end
	elseif diagonalIn and not diagonalOut then
		--print(diagonalIn.Instance.CollisionGroup)
		self:AlignCharacter(deltaTime, diagonalIn)
	elseif sideways then
		--print(sideways.Instance.CollisionGroup)
		self:AlignCharacter(deltaTime, sideways)
	elseif centerFwd then
		--print(centerFwd.Instance.CollisionGroup)
		self:AlignCharacter(deltaTime, centerFwd)
	end

	if input.Magnitude > 1 then input = input.Unit end
	self.ClimbForce.VectorVelocity = (rVector * input.X + uVector * input.Y) * CLIMB_SPEED
end

function MovementController:AttemptClimb(ray: RaycastResult): boolean
	if self.Climbing then return end
	if self.Character:GetAttribute("Ragdolled") then return end

	if self.Dashing then self:DashStop() end
	if self.Running then self:RunStop() end

	local hasMana = MagicController:HasMana()
	if not hasMana then return end
	self:ClimbStart(ray)
end

function MovementController:_run(baseSpeed: number)
	AnimationController:Play("Run")
	self.Character.Humanoid.WalkSpeed = baseSpeed * 1.5
end

function MovementController:_manaRun(baseSpeed: number)
	AnimationController:Play("ManaRun")
	self.RunTrail.Enabled = true
	self.Character.Humanoid.WalkSpeed = baseSpeed * 2
end

function MovementController:RunStart(hasMana: boolean)
	if PRINT_STARTS then print('start \t\trun') end

	local statChanges = Knit.Player.Data.StatChanges
	local speedBoost = statChanges.SpeedBoost.Value

	local BASE_PLAYER_SPEED = BASE_WALK_SPEED + speedBoost

	if hasMana then
		MagicController:ChargeMana(false)
		self:_manaRun(BASE_PLAYER_SPEED)
	else
		self:_run(BASE_PLAYER_SPEED)
	end

	self.Running = true
	ActionService.ActionStarted:Fire("run")
end

function MovementController:RunStop()
	if PRINT_STOPS then print('stop \t\trun') end

	local humanoid = self.Character:FindFirstChild("Humanoid")
	local statChanges = Knit.Player.Data.StatChanges
	local speedBoost = statChanges.SpeedBoost.Value
	local BASE_PLAYER_SPEED = BASE_WALK_SPEED + speedBoost

	humanoid.WalkSpeed = BASE_PLAYER_SPEED
	self.RunTrail.Enabled = false
	AnimationController:Stop("ManaRun")
	AnimationController:Stop("Run")

	self.Running = false
	ActionService.ActionStopped:Fire("run")
end

function MovementController:AttemptRun()
	if self.Climbing or self.Running or self.Dashing then return end
	if self.Character:GetAttribute("Ragdolled") then return end

	local hasMana = MagicController:HasMana()
	self:RunStart(hasMana)
end

function MovementController:DashStart(hasMana: boolean)
	if PRINT_STARTS then print('start \t\tdash') end

	local hrp = self.Character:FindFirstChild("HumanoidRootPart")
	self._dashTrove = self._trove:Extend()

	local direction = ""
	if UserInputService:IsKeyDown(Enum.KeyCode.W) then
		self._DashDirection = 0
		direction = "Forward"
	elseif UserInputService:IsKeyDown(Enum.KeyCode.A) then
		self._DashDirection = 90
		direction = "Left"
	elseif UserInputService:IsKeyDown(Enum.KeyCode.D) then
		self._DashDirection = -90
		direction = "Right"
	else
		self._DashDirection = 180
		direction = "Backward"
	end

	self._dashVelocity = self._dashTrove:Add(Instance.new("BodyVelocity"))
	self._dashVelocity.Parent = hrp
	self._dashVelocity.MaxForce = Vector3.new(10e9, 0, 10e9)
	self._dashVelocity.P = 1250
	self._dashVelocity.Velocity = (hrp.CFrame * CFrame.Angles(0, math.rad(self._DashDirection), 0)).lookVector * (hasMana and 60 or 50)

	AnimationController:Play("Dash"..direction)

	if hasMana then
		MagicController:ChargeMana(false)
		self.ManaDashSound:Play()
		self.DashParticles.Enabled = true
	end

	self.Dashing = true
	task.delay(
		DASH_DURATION,
		function()
			self:DashStop()
		end
	)
	self._dashTrove:BindToRenderStep("dash_update", Enum.RenderPriority.Character.Value, function(deltaTime)
		self:DashUpdate(deltaTime)
	end)

	ActionService.ActionStarted:Fire("dash")
end

function MovementController:DashStop()
	if PRINT_STOPS then print('stop \t\tdash') end

	self.DashParticles.Enabled = false

	self.Dashing = false
	self._dashTrove:Destroy()
	ActionService.ActionStopped:Fire("dash")
end

function MovementController:DashUpdate()
	local hrp = self.Character:FindFirstChild("HumanoidRootPart")

	if self._dashVelocity and self.Dashing then
		hrp.CFrame = CFrame.new(hrp.Position, hrp.Position + (workspace.CurrentCamera.CFrame.LookVector * Vector3.new(1, 0, 1)))
		self._dashVelocity.Velocity = (hrp.CFrame * CFrame.Angles(0, math.rad(self._DashDirection), 0)).lookVector * math.round(self._dashVelocity.Velocity.Magnitude)
	else
		self:DashStop()
	end
end

function MovementController:AttemptDash()
	if self.Climbing or self.Dashing then return end
	if self.Character:GetAttribute("Ragdolled") then return end

	if self.Running then self:RunStop() end

	local hasMana = MagicController:HasMana()
	self:DashStart(hasMana)
end

function MovementController:HandleJump()
	local hrp = self.Character:FindFirstChild("HumanoidRootPart")
	local hmnd = self.Character:FindFirstChild("Humanoid")

	if self.Climbing then if PRINT_DIAGNOSES then print("cancel climb") end self:ClimbStop() return end

	local forwardOrigin = hrp.Position
	local forwardRay = workspace:Raycast(forwardOrigin, hrp.CFrame.LookVector * Vector3.new(2, 0, 2), raycastParams)
	if VISUALIZE then CastVisualizer.visualize(forwardOrigin, forwardOrigin + hrp.CFrame.LookVector * Vector3.new(2, 0, 2)) end

	if forwardRay and hmnd.FloorMaterial == Enum.Material.Air then
		local forwardOrigin = hrp.Position
		local forwardRay = workspace:Raycast(forwardOrigin, hrp.CFrame.LookVector * Vector3.new(2, 0, 2), raycastParams)

		if forwardRay and forwardRay.Instance.Anchored and forwardRay.Instance.CanCollide and not forwardRay.Instance:IsA("TrussPart") then
			self:AttemptClimb(forwardRay)
			return
		end
	end
end

function MovementController:OnManaEmptied()
	if self.Running then
		local statChanges = Knit.Player.Data.StatChanges
		local speedBoost = statChanges.SpeedBoost.Value

		local BASE_PLAYER_SPEED = BASE_WALK_SPEED + speedBoost

		self.RunTrail.Enabled = false
		AnimationController:Stop("ManaRun")
		self:_run(BASE_PLAYER_SPEED)
	end

	if self.Climbing then
		--print("mana emptied")
		self:ClimbStop()
	end
end

function MovementController:OnManaColorChanged(color: Color3)
	while not Knit.Player:GetAttribute("IdentityLoaded") or not self.RunTrail or not self.DashParticles do
		RunService.Heartbeat:Wait()
	end

	self.RunTrail.Color = ColorSequence.new(color)
	self.DashParticles.Color = ColorSequence.new(color)
end

function MovementController:OnCharacterAdded(character: Model)
	self.Character = character

	self.Climbing = false

	self.Running = false

	self.RunTrail = self._trove:Clone(VFX.ManaRunTrail)
	self.RunTrail.Parent = character:WaitForChild("Torso")
	self.RunTrail.Attachment0 = character.Torso.BodyFrontAttachment
	self.RunTrail.Attachment1 = character.Torso.BodyBackAttachment
	self.RunTrail.Enabled = false

	self.Dashing = false

	self.DashParticles = self._trove:Clone(VFX.ManaStopParticle)
	self.DashParticles.Parent = character.Torso

	self.ManaDashSound.Parent = character:WaitForChild("HumanoidRootPart")

	raycastParams:AddToFilter(character)
end

function MovementController:OnPlayerAdded(player)
	player.CharacterAdded:Wait()
	local character = player.Character
	raycastParams:AddToFilter(character)
	self._trove:Connect(player.CharacterAdded, function(character)
		raycastParams:AddToFilter(character)
	end)
end

function MovementController:OnPlayerRemoving(player)
	if player == Knit.Player then
		self:Destroy()
	end
end

function MovementController:KnitInit()
	self._trove = Trove.new()

	self.ManaDashSound = self._trove:Clone(SFX.ManaDash)

	for _,player in game.Players:GetPlayers() do
		coroutine.wrap(function() self:OnPlayerAdded(player) end)()
	end
	self._trove:Connect(game.Players.PlayerAdded, function(...) self:OnPlayerAdded(...) end)
end

function MovementController:KnitStart()
	ActionService = Knit.GetService("ActionService")
	IdentityService = Knit.GetService("IdentityService")
	ManaService = Knit.GetService("ManaService")

	AnimationController = Knit.GetController("AnimationController")
	InputController = Knit.GetController("InputController")
	MagicController = Knit.GetController("MagicController")

	if Knit.Player.Character then self:OnCharacterAdded() end
	self._trove:Connect(Knit.Player.CharacterAdded, function(...) self:OnCharacterAdded(...) end)

	self._trove:Connect(IdentityService.ManaColorChanged, function(...) self:OnManaColorChanged(...) end)
	self._trove:Connect(ManaService.ManaEmptied, function(...) self:OnManaEmptied(...) end)

	if PRINT_STARTS then print("MovementController started") end
end

function MovementController:Destroy()
	self._trove:Destroy()
end

return MovementController

File: src/client/modules/Controllers/flamework/movement-controller.ts
import { Controller, Dependency, OnStart } from "@flamework/core";
import { ReplicatedStorage, TweenService, UserInputService, Workspace } from "@rbxts/services";
import { OnLocalCharacterAdded } from "../../../../../types/lifecycles";
import { Components } from "@flamework/components";
import { CharacterClient as Character } from "client/modules/components/character-client";
import { Trove } from "@rbxts/trove";
import { InputController } from "./input-controller";
import { hasLineOfSight } from "shared/modules/line-of-sight";
import { Events } from "client/modules/networking";

const BASE_CLIMB_SPEED = 10;
const BASE_WALK_SPEED = 20;
const DASH_DURATION = 0.4;
const TRAINED_CLIMB_BONUS_DURATION = 10;

const VFX = ReplicatedStorage.Effects.Visuals;
const SFX = ReplicatedStorage.Effects.Sounds;

export type Direction = "forward" | "backward" | "left" | "right";

const directionToAngle: {[direction: string]: number} = {
	forward: 0,
	backward: 180,
	left: 90,
	right: -90
}

@Controller()
export class MovementController implements OnStart, OnLocalCharacterAdded {
	private raycastParams = new RaycastParams();
	private character: Character | undefined;

	private climbTrove = new Trove();
	private climbForce: LinearVelocity | undefined;
	private climbConstraint: AlignOrientation | undefined;
	private goalPart: Part | undefined;
	private goalAttachment: Attachment | undefined;

	private dashTrove = new Trove();
	private dashVelocity: BodyVelocity | undefined;

	isRunning = false;
	isClimbing = false;
	isDodging = false;

	onStart(): void {
		this.raycastParams.CollisionGroup = "Characters";
		this.raycastParams.FilterType = Enum.RaycastFilterType.Exclude;
		this.raycastParams.IgnoreWater = true;

		Events.manaEvents.manaEmptied.connect(() => this.onManaEmptied());
	}

	onLocalCharacterAdded(character: Model): void {
		const components = Dependency<Components>();
		components.waitForComponent<Character>(character).andThen((value) => this.character = value);
		this.raycastParams.AddToFilter(character);
	}

	onManaEmptied(): void {
		if (this.isRunning) {
			// downgrade from manarun to normal run
		} else if (this.isClimbing) {
			this.stopClimb();
		}
	}

	private alignCharacterToWall(deltaTime: number, wallCastResult: RaycastResult): void {
		if (!this.goalPart) return;
		const goal = CFrame.lookAt(wallCastResult.Position.add(wallCastResult.Normal), wallCastResult.Position)
		this.goalPart.CFrame = goal;

		if (!this.character) return;
		const humanoidRootPart = this.character.getHumanoidRootPart();
		humanoidRootPart.CFrame = humanoidRootPart.CFrame.Lerp(goal, 1-math.pow(1e-9, deltaTime));
	}

	private newClimbForce(): LinearVelocity {
		const force = new Instance("LinearVelocity");
		force.MaxForce = math.huge;
		return force;
	}

	private newClimbConstraint(): AlignOrientation {
		const constraint = new Instance("AlignOrientation");
		constraint.AlignType = Enum.AlignType.Parallel;
		constraint.MaxTorque = 1e7;
		return constraint;
	}

	private newGoalPart(): Part {
		const part = new Instance("Part");
		part.Size = new Vector3(4, 1.2, 2);
		part.CanCollide = false;
		part.CanQuery = false;
		part.CanTouch = false;
		part.Anchored = true;
		part.Transparency = 1;
		return part;
	}

	private newDashVelocity(): BodyVelocity {
		const bv = new Instance("BodyVelocity");
		bv.MaxForce = new Vector3(1e10, 0, 1e10);
		bv.P = 1250;
		return bv;
	}

	startClimb(wallCastResult: RaycastResult) {
		if (!this.character) return;

		this.isClimbing = true;

		const characterRootPart = this.character.getHumanoidRootPart();

		this.goalPart = this.climbTrove.add(this.newGoalPart());
		this.goalPart.Parent = characterRootPart

		this.goalAttachment = this.climbTrove.add(new Instance("Attachment"));
		this.goalAttachment.Parent = this.goalPart;

		this.climbForce = this.climbTrove.add(this.newClimbForce());
		this.climbForce.Parent = this.character.instance;
		this.climbForce.Attachment0 = characterRootPart.RootAttachment;
		
		this.climbConstraint = this.climbTrove.add(this.newClimbConstraint());
		this.climbConstraint.Parent = this.character.instance;
		this.climbConstraint.Attachment0 = characterRootPart.RootAttachment;
		this.climbConstraint.Attachment1 = this.goalAttachment;

		const humanoid = this.character.instance.Humanoid;

		humanoid.AutoRotate = false;
		humanoid.SetStateEnabled(Enum.HumanoidStateType.Freefall, false)
		humanoid.SetStateEnabled(Enum.HumanoidStateType.Running, false)
		humanoid.ChangeState(Enum.HumanoidStateType.Climbing);

		this.alignCharacterToWall(0, wallCastResult);

		this.climbTrove.bindToRenderStep(
			"climb_update",
			Enum.RenderPriority.Character.Value,
			(deltaTime) => {
				const input = InputController.inputVector;
				this.updateClimb(deltaTime, input);
			}
		);
	}

	climbUp(goal: CFrame) {
		if (!this.character) return;
		
		const humanoidRootPart = this.character.getHumanoidRootPart();
		
		humanoidRootPart.Anchored = true;
		const lookVector = humanoidRootPart.CFrame.LookVector;
		const upVector = humanoidRootPart.CFrame.UpVector;
		
		goal = goal.add(lookVector.mul(0.5).add(upVector));
		const tween = TweenService.Create(
			humanoidRootPart,
			new TweenInfo(0.3, Enum.EasingStyle.Linear),
			{ CFrame: goal }
		)
		tween.Completed.Once(() => {
			humanoidRootPart.Anchored = false;
			this.stopClimb();
		})
		tween.Play();
	}

	private calculateClimbSpeed(mana: number, climbTrains: number, climbBoost: number) {
		return (-1e-3 * mana * (mana-200) + BASE_CLIMB_SPEED + climbTrains/TRAINED_CLIMB_BONUS_DURATION) * (1+climbBoost)
	}

	updateClimb(deltaTime: number, inputVector: Vector2): void {
		if (!this.character) return;
		
		const humanoidRootPart = this.character.getHumanoidRootPart();
		const upVector = humanoidRootPart.CFrame.UpVector;

		const floorCheck = Workspace.Raycast(
			humanoidRootPart.Position.sub(Vector3.yAxis.mul(0.5)),
			upVector.mul(-1),
			this.raycastParams
		);
		if (floorCheck) {
			this.stopClimb();
			return;
		}
		
		const lookVector = humanoidRootPart.CFrame.LookVector;
		const rightVector = humanoidRootPart.CFrame.RightVector;
		
		const doubledLookVector = lookVector.mul(2);
		const withInput = rightVector.mul(inputVector.X);

		let diagonalIn, diagonalOut, sideways;
		if (inputVector.X !== 0) {
			const againstInput = withInput.mul(-1);
			
			const inOrigin = humanoidRootPart.Position.add(withInput);
			const inPoint = humanoidRootPart.Position.add(againstInput).add(doubledLookVector);
			const outOrigin = humanoidRootPart.Position.add(againstInput);
			const outPoint = humanoidRootPart.Position.add(withInput).add(doubledLookVector);

			diagonalIn = Workspace.Raycast(inOrigin, inPoint.sub(inOrigin), this.raycastParams);
			diagonalOut = Workspace.Raycast(outOrigin, outPoint.sub(outOrigin), this.raycastParams);
			sideways = Workspace.Raycast(humanoidRootPart.Position, withInput.mul(2), this.raycastParams);
		}

		const blockSize = new Vector3(humanoidRootPart.Size.X, humanoidRootPart.Size.Y/2, 0.1);
		const blockForwardCast = Workspace.Blockcast(humanoidRootPart.CFrame, blockSize, doubledLookVector, this.raycastParams);

		if (!(blockForwardCast || diagonalIn || diagonalOut)) {
			this.stopClimb();
			return;
		}

		let seesTop = false;
		const headCastOrigin =
			humanoidRootPart.CFrame
				.add(upVector.mul(1.5))
				.sub(lookVector.mul(0.5));
		const headCastSize = new Vector3(2, 0.1, 0.1);
		if (!Workspace.Blockcast(headCastOrigin, headCastSize, doubledLookVector, this.raycastParams)) {
			const to = humanoidRootPart.Position
				.add(lookVector.mul(1.5))
				.add(Vector3.yAxis);
			seesTop = hasLineOfSight(headCastOrigin.Position, to, this.raycastParams);
		}

		const centerForward = Workspace.Raycast(humanoidRootPart.Position, doubledLookVector, this.raycastParams);

		if (seesTop) {
			const ledgeCastOrigin = humanoidRootPart.CFrame
				.mul(CFrame.Angles(math.rad(-10), 0, 0))
				.add(Vector3.yAxis);
			const ledgeCastSize = new Vector3(0.01, 2, 0.01);
			const ledgeCast = Workspace.Blockcast(ledgeCastOrigin, ledgeCastSize, lookVector, this.raycastParams);

			if (ledgeCast)
				this.climbUp(new CFrame(ledgeCast.Position).mul(humanoidRootPart.CFrame.Rotation));
			else
				this.stopClimb();
		} else if (diagonalIn && !diagonalOut)
			this.alignCharacterToWall(deltaTime, diagonalIn);
		else if (sideways)
			this.alignCharacterToWall(deltaTime, sideways);
		else if (centerForward)
			this.alignCharacterToWall(deltaTime, centerForward);

		if (this.climbForce) {
			const lateral = rightVector.mul(inputVector.X)
			const vertical = upVector.mul(inputVector.Y)
			let combined = lateral.add(vertical);
			if (combined.Magnitude > 1) combined = combined.Unit;
			this.climbForce.VectorVelocity = combined
				.mul(this.calculateClimbSpeed(100, 0, 0)); // TODO: fix this
		}
	}

	stopClimb(): void {
		if (!this.character) return;
		
		this.isClimbing = false;
		this.climbTrove.clean();

		const humanoid = this.character.instance.Humanoid;
		humanoid.AutoRotate = true;
		humanoid.SetStateEnabled(Enum.HumanoidStateType.Freefall, true)
		humanoid.SetStateEnabled(Enum.HumanoidStateType.Running, true)
	}

	startRun(): void {
		if (this.isClimbing || this.isRunning || this.isDodging) return;
		if (!this.character) return;
		if (this.character.instance.GetAttribute("isRagdolled")) return;

		this.isRunning = true;

		const BASE_PLAYER_SPEED = BASE_WALK_SPEED; // + speedBoost
		this.character.instance.Humanoid.WalkSpeed = BASE_PLAYER_SPEED * 1.5;
	}

	stopRun(): void {
		if (!this.character) return;
		if (!this.isRunning) return;

		this.isRunning = false;

		const BASE_PLAYER_SPEED = BASE_WALK_SPEED; // + speedBoost
		this.character.instance.Humanoid.WalkSpeed = BASE_PLAYER_SPEED;
	}

	startDodge(hasMana: boolean, direction: Direction): void {
		if (this.isClimbing || this.isDodging) return;
		if (!this.character) return;
		if (this.character.instance.GetAttribute("isRagdolled")) return;

		if (this.isRunning) this.stopRun();

		this.isDodging = true;

		let directionAngle = directionToAngle[direction];
		const humanoidRootPart = this.character.getHumanoidRootPart();

		this.dashVelocity = this.dashTrove.add(this.newDashVelocity());
		this.dashVelocity.Parent = humanoidRootPart;
		this.dashVelocity.Velocity = humanoidRootPart.CFrame
			.mul(CFrame.Angles(0, math.rad(directionAngle), 0))
			.LookVector
			.mul(hasMana ? 60 : 50);

		task.delay(DASH_DURATION, () => this.stopDodge());
		
		this.dashTrove.bindToRenderStep(
			"dash_update",
			Enum.RenderPriority.Character.Value,
			() => this.updateDodge(directionAngle)
		);
	}

	updateDodge(angle: number): void {
		if (!this.character) return;
		if (!Workspace.CurrentCamera) return;

		const humanoidRootPart = this.character.getHumanoidRootPart();
		if (this.dashVelocity && this.isDodging) {
			humanoidRootPart.CFrame = new CFrame(
				humanoidRootPart.Position,
				humanoidRootPart.Position
					.add(
						Workspace.CurrentCamera.CFrame.LookVector
						.mul(new Vector3(1, 0, 1))
					)
				)
			const speed = math.round(this.dashVelocity.Velocity.Magnitude)
			this.dashVelocity.Velocity = humanoidRootPart.CFrame
				.mul(CFrame.Angles(0, math.rad(angle), 0))
				.LookVector
				.mul(speed);
		} else {
			this.stopDodge();
		}
	}

	stopDodge(): void {
		this.isDodging = false;
		this.dashTrove.clean();
	}

	handleJump(): void {
		if (!this.character) return;

		if (this.isClimbing) {
			this.stopClimb();
			return;
		}

		const humanoidRootPart = this.character?.getHumanoidRootPart();
		const forwardCast = Workspace.Raycast(
			humanoidRootPart.Position,
			humanoidRootPart.CFrame.LookVector.mul(new Vector3(2, 0, 2)),
			this.raycastParams
		);

		const inAir = this.character.instance.Humanoid.FloorMaterial === Enum.Material.Air;
		if (forwardCast && inAir) {
			const castInstance = forwardCast.Instance;
			if (castInstance.Anchored &&
				castInstance.CanCollide &&
				!castInstance.IsA("TrussPart")
			) {
				this.startClimb(forwardCast);
			}
		}
	}
}